presto-main/src/main/java/io/prestosql/connector/informationschema/InformationSchemaPageSource.java

```diff
@@ -346,13 +355,35 @@ private void addRolesRecords()
         }
     }
 
+    private void addRoleAuthorizationDescriptorRecords()
+    {
+        try {
+            accessControl.checkCanShowRoles(session.toSecurityContext(), catalogName);
+        }
+        catch (AccessDeniedException exception) {
+            return;
+        }
+
+        for (RoleGrant grant : metadata.listAllRoleGrants(session, catalogName, roles, grantees, limit)) {
```

Comment:
Do you think it is doable to implement predicate and limit pushdown only here (instead of `SqlStandardAccessControlMetadata`)? That way it will out of the box for any other connectors or even for any access control metadata in hive. That would possibly require adding different methods to `metadata` (more granular ones).

WDYT?

Reply:
That's what I had first, but @electrum correctly pointed out that this is a vert Hive specific.
So in this design here we just pass the predicates down, and then let the Connectors do what's best for them. (For example only HMS has this weird limit that the only APIs are getting Roles by a single Principal, or Principals by a single Role).

Reply:
I see, but I am afraid that we went to far with this being too much hive specific and so it can be difficult in future to reuse it as one might need to reimplement predicate pushdown in the one use case.

> HMS has this weird limit that the only APIs are getting Roles by a single Principal, or Principals by a single Role

Could expose this in Connector metadata SPI? I would imagine that in future we could also have batch version of these methods. In case when batch version is not implemented (like in Hive) we could fallback to using methods to access roles for given principal or principals for given role.

WDYT?

Reply:
ConnectorMetadata has now two relevant methods: (1) list all roles, and (2) list all role grants (with optional filter criteria).

I think connectors will be quite different on this front. Databases (Postgres, Oracel, MySQL, Phoenix, etc) have their own `information_schema` and the connector would just turn this into a query on *their* `information_schema` in that case we'd want to pass the predicate through. 
Others (like Hive) have limited and different APIs, and that needs some logic to be "optimal". Yet, other connectors have other way to query their role metadata.

Both methods may or may not be batched, but it's completely hidden now. The connector decides how to implement them. Note that the Hive implementation is completely hidden in the Hive Metadata, and the only part the InformationSchema needs are these two methods to be called. Everything else is automated (extracting the predicate, passing it down, etc)

If we do not want to leak Hive specifics into the InformationSchemas code I do not see how to do it fundamentally differently. If I put specific logic any higher than ConnectorMetadata we'd leak the details of Hive up.

But maybe I don't understand what you mean...
Perhaps could outline what methods you would add to ConnectorMetadata, and how would the code in InformationSchemaPageSource use them...?

Thanks!

Reply:
Thanks!